scipy/stats/_morestats.py

```diff
@@ -4019,6 +4030,43 @@ def wilcoxon(x, y=None, zero_method="wilcox", correction=False,
     of ranks of positive differences) whereas it is 24 in the two-sided
     case (the minimum of sum of ranks above and below zero).
 
+    In the example above, the differences in height between paired plants are
+    provided to `wilcoxon` directly. Alternatively, `wilcoxon` accepts two
+    samples of equal length, calculates the differences between paired
+    elements, then performs the test. Consider the samples ``x`` and ``y``:
+
+    >>> import numpy as np
+    >>> x = np.array([0.5, 0.825, 0.375, 0.5])
+    >>> y = np.array([0.525, 0.775, 0.325, 0.55])
+    >>> res = wilcoxon(x, y, alternative='greater')
+    >>> res
+    WilcoxonResult(statistic=5.0, pvalue=0.5625)
+
+    Note that had we calculated the differences by hand, the test would have
+    produced different results:
+
+    >>> d = [-0.025, 0.05, 0.05, -0.05]
+    >>> ref = wilcoxon(d, alternative='greater')
+    >>> ref
+    WilcoxonResult(statistic=6.0, pvalue=0.4375)
+
+    The substantial difference is due to roundoff error in the results of
+    ``x-y``:
+
+    >>> d - (x-y)
+    array([2.08166817e-17, 6.93889390e-17, 1.38777878e-17, 4.16333634e-17])
+
+    Even though we expected all the elements of ``(x-y)[1:]`` to have the same
+    magnitude ``0.05``, they have slightly different magnitudes in practice,
+    and therefore are assigned different ranks in the test. Before performing
+    the test, consider calculating ``d`` and adjusting it as necessary to
+    ensure that theoretically identically values are not numerically distinct.
+    For example:
+
+    >>> d2 = np.around(x - y, decimals=3)
```

Comment:
Can we then not round directly in the test?

Reply:
Fixing it for the user was one of my three suggestions in gh-17667. I chose to take the doc-only route because *just* modifying the user's input is not an option, IMO. It's not always the correct thing to do. Something more needs to be done:
- expose a tolerance parameter so the user can turn this on or off (but I don't really think this deserves a new parameter)
- we could have a fixed tolerance and warn if the data gets modified
- we could have fixed tolerance detect the problem, but just warn without modifying the data

In any of those cases, we'd need some subset of the ideas in this PR to document what is going on.

Reply:
Can we not use machine precision here? Like this is working fine `wilcoxon(np.around(x - y, decimals=15), alternative='greater')`

Reply:
We can use any number 3-15 in this case. I chose 3 because the reader expects the results to have three significant decimal digits.

Reply:
What I mean is that we could replace in the code of the function the line `d = x - y` by `d = np.around(x - y, decimals=15)`. This way we don't really modify the users' input as we are doing things up to machine resolution.

Reply:
It's the most obvious and easiest thing, but I would be uncomfortable changing the existing function to do that : / Can we get some other opinions?

Reply:
@chrisb83 @tirthasheshpatel @steppi roundoff errors in `x - y` can dramatically affect the results of `wilcoxon` because it's a rank-based test. This PR would add a warning and show an example of how to deal with this: calculate `d = x - y`, round as needed, and pass `d` into `wilcoxon`. However, what do you think about rounding to `np.finfo(dtype).resolution` inside `wilcoxon` automatically? (I am hesitant to take action that _might_ do harm even though I think it would almost certainly do more good than harm.)

Reply:
I lean towards not modifying the data. The choice of how to discretize should depend on the sensitivity and precision of the measuring device and the scale of what's being measured. The person running the test should always think about these things, and I think it's fine just to document the problem and warn users.

I've thought of two examples where we wouldn't want to enforce rounding:

First, here's a contrived example.

Suppose we have two nano-scale measurement devices which can measure weights of molecules down to `1e-18` grams (this is contrived because in reality these weights would be reported in daltons or something.) We want to test if one of the devices has a systematic bias compared to the other. We find 5 molecules and measure them with each device to get
```python
m1 = np.array([3.21e-16, 1.04e-16, 0.97e-16, 7.25e-16, 0.86e-16])
m2 = np.array([3.19e-16, 1.05e-16, 0.94e-16, 7.23e-16, 0.81e-16])
```

Then 
```python
np.around(m1 - m2, decimals=15)
```
yields `np.array([0., -0., 0., 0., 0.])`

but the machine is able to subtract floating point numbers at this resolution. What matters isn't the number of digits after the decimal place, but the relative distance between the numbers.

Again, I know this example is contrived, but I still don't feel confident saying that no one anywhere is working with measurements this small in magnitude.

Second, someone might be interesting in investigating how the wilcoxon test behaves under random perturbations. They're aware that round-off errors can have a large impact on the results and want to test how large this impact can be for a problem of interest. This is a pretty niche use-case, but I wouldn't want to deny this possibility.

Thoughts?

Reply:
Interesting, reading your comment I am thinking we could then scale (remove the mean here) to get get the relative difference without the noice of the actual scale.

In any case a warning seems wanted. But I would say this is pretty general and we could find some global place to put this in. We have reports like that from time to time.

Reply:
Both `m1` and `m2` would need to be rescaled by the same value, otherwise we'll change the results of the test, possibly catastrophically.

Consider
```python
m1 = np.array([0.9, 1., 1.1, 1.2, 2.])
m2 = np.array([1.8, 2.0, 2.2, 2.4, 7.])
```

then

```python
m1 - m2
# array([-0.9, -1. , -1.1, -1.2, -5. ])
```

but

```python
m1 / np.mean(m1) - m2 / np.mean(m2) 
#  array([ 0.14139087,  0.15710096,  0.17281106,  0.18852116, -0.65982405])
```

Something I considered was was to divide out by the mean of the floating point exponents of all of the numbers in `m1` and `m2`.

Something like


```python
def wilcoxon_normalize(m1, m2):
    mean_exp = np.mean([np.frexp(m1)[1], np.frexp(m2)[1])
    return np.around(m1*2**(-mean_exp) - m*2**(-mean_exp)), decimals=15)
```

but I think I'd still prefer just to leave rounding up to the user and maybe just add a tutorial on rounding for the wilcoxon test.

Reply:
Makes sense, thank you 👍 

So are you saying this should be a tutorial instead of the change proposed in this PR? Or should we first go with that?

Reply:
ok then let's get this in and we can come back later with a plan.